lib/connectors/memory.js

```diff
@@ -250,6 +250,46 @@ Memory.prototype.updateOrCreate = function(model, data, options, callback) {
     }
   });
 };
+Memory.prototype.patchOrCreateWithWhere =
+  Memory.prototype.upsertWithWhere = function(model, where, data, options, callback) {
+    var self = this;
+    var element = Object.keys(where)[0];
```

Comment:
I agree, this is suspicious and most likely will make it difficult to extend `upsertWithWhere` in the future.

My expectation is that inside the connector, we should simply call `all` (or `find`) with `where` filter which was passed in from DAO. If there should be any (artificial/temporary) restrictions on what is allowed in the `where` filter, then that should be handled in DAO.

As for `data[element] = where[element]`, I would expect that the caller of the API is required to provide full `data` object with changes to apply, so that we don't have to parse possibly arbitrary `where` clause here.